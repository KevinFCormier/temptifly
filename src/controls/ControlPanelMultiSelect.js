'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { MultiSelect } from 'carbon-components-react'
import Tooltip from '../components/Tooltip'

class ControlPanelMultiSelect extends React.Component {
  static propTypes = {
    control: PropTypes.object,
    controlId: PropTypes.string,
    handleChange: PropTypes.func,
    i18n: PropTypes.string
  };

  constructor(props) {
    super(props)
    this.state = {}
  }

  setControlRef = (control, ref) => {
    this.multiSelect = control.ref = ref
  };

  render() {
    const { controlId, i18n, control } = this.props
    const { name, placeholder: ph = '' } = control

    // see if we need to add user additions to available (from editing the yaml file)
    const { userData, userMap, hasCapturedUserSource } = control
    let { active = [], available, availableMap } = control
    if (userData) {
      if (!hasCapturedUserSource) {
        available = [...userData, ...available]
        availableMap = { ...userMap, ...availableMap }
      } else {
        // if user edited the source, we can't automatically update it
        active = available = [i18n('creation.view.policy.custom')]
        availableMap = undefined
      }
    }
    if (!Array.isArray(active)) {
      active=[]
    }

    // place holder
    let placeholder = ph
    if (active.length > 0) {
      const activeKeys = []
      active.forEach(k => {
        if (typeof availableMap === 'object' && availableMap[k]) {
          const { name: n } = availableMap[k]
          activeKeys.push(n || k)
        } else {
          activeKeys.push(k)
        }
      })
      placeholder = activeKeys.join(', ')
    }

    // change key if active changes so that carbon component is re-created with new initial values
    const key = `${controlId}-${active.join('-')}`
    return (
      <React.Fragment>
        <div
          className="creation-view-controls-multiselect"
          ref={this.setControlRef.bind(this, control)}
        >
          <label
            className="creation-view-controls-multiselect-title"
            htmlFor={controlId}
          >
            {name}
            <Tooltip control={control} i18n={i18n} />
          </label>
          <MultiSelect.Filterable
            key={key}
            id={controlId}
            items={available}
            initialSelectedItems={active}
            placeholder={placeholder}
            itemToString={item => item}
            sortItems={items => items}
            onChange={this.handleSelectionChange.bind(this)}
          />
        </div>
      </React.Fragment>
    )
  }

  handleSelectionChange(evt) {
    const { control } = this.props
    const { isOneSelection } = control
    if (isOneSelection) {
      // close on one selection
      this.handleChange(evt)
    } else {
      // close when user clicks outside of menu
      // unfortunately MultiSelect.Filterable doesn't have an onClose
      this.multiSelect.selectedItems = evt.selectedItems
      const menu = this.multiSelect.getElementsByClassName(
        'bx--list-box__menu'
      )
      if (menu && menu.length > 0) {
        if (!this.multiSelect.observer) {
          this.multiSelect.observer = new MutationObserver(() => {
            this.handleChange({
              selectedItems: this.multiSelect.selectedItems
            })
            this.multiSelect.observer.disconnect()
            delete this.multiSelect.observer
          })
          this.multiSelect.observer.observe(menu[0].parentNode, {
            childList: true
          })
        }
      } else if (!this.multiSelect.observer) {
        this.handleChange({ selectedItems: this.multiSelect.selectedItems })
      }
    }
  }

  handleChange(evt) {
    const { control, handleChange } = this.props
    control.active = evt.selectedItems
    handleChange(evt)
  }
}

export default ControlPanelMultiSelect
