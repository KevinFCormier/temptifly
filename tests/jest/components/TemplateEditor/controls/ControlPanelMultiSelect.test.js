'use strict'

import React from 'react'
import ControlPanelMultiSelect from '../../../../../src/controls/ControlPanelMultiSelect'
import renderer from 'react-test-renderer'

import { control } from '../../TestingData'

describe('ControlPanelMultiSelect component', () => {
  it('renders as expected', () => {
    const fn = jest.fn()
    const component = renderer.create(
      <ControlPanelMultiSelect
        key={'key'}
        control={control}
        controlId={'controlId'}
        handleChange={fn}
        i18n={fn}
      />
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})
