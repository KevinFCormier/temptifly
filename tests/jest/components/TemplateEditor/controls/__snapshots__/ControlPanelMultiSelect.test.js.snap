// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPanelMultiSelect component renders as expected 1`] = `
<div
  className="creation-view-controls-multiselect"
>
  <label
    className="creation-view-controls-multiselect-title"
    htmlFor="controlId"
  >
    creation.app.name
    <div
      className="bx--tooltip__label"
    >
      
      <div
        aria-expanded={false}
        aria-haspopup="true"
        aria-label="tooltip"
        className="bx--tooltip__trigger"
        id=""
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        role="button"
        tabIndex={0}
        title=""
      >
        <svg
          alt="tooltip"
          aria-label="tooltip"
          className={undefined}
          fill={undefined}
          fillRule="evenodd"
          height="16"
          name={undefined}
          role="img"
          style={undefined}
          viewBox="0 0 16 16"
          width="16"
        >
          <title>
            
          </title>
          <path
            d="M8,14.5 C11.5898509,14.5 14.5,11.5898509 14.5,8 C14.5,4.41014913 11.5898509,1.5 8,1.5 C4.41014913,1.5 1.5,4.41014913 1.5,8 C1.5,11.5898509 4.41014913,14.5 8,14.5 Z M8,16 C3.581722,16 0,12.418278 0,8 C0,3.581722 3.581722,0 8,0 C12.418278,0 16,3.581722 16,8 C16,12.418278 12.418278,16 8,16 Z"
          />
          <polygon
            points="9 13 7 13 7 7 9 7"
          />
          <circle
            cx="8"
            cy="4"
            r="1"
          />
        </svg>
      </div>
    </div>
  </label>
  <div
    aria-invalid={undefined}
    aria-label="Choose an item"
    className="bx--multi-select bx--combo-box bx--list-box"
    data-invalid={undefined}
    onClick={[Function]}
    onKeyDown={[Function]}
    role="listbox"
    tabIndex="-1"
  >
    <div
      aria-controls={null}
      aria-expanded={false}
      aria-haspopup={true}
      aria-label="open menu"
      aria-owns={null}
      className="bx--list-box__field"
      data-toggle={true}
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex="0"
      type="button"
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-expanded={false}
        autoComplete="off"
        className="bx--text-input"
        disabled={false}
        id="controlId"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder=""
        role="combobox"
        value=""
      />
      
      <div
        className="bx--list-box__menu-icon"
      >
        <svg
          alt="Open menu"
          aria-label="Open menu"
          className={undefined}
          fill={undefined}
          fillRule="evenodd"
          focusable="false"
          height="5"
          name={undefined}
          role="img"
          style={undefined}
          viewBox="0 0 10 5"
          width="10"
        >
          <title>
            Open menu
          </title>
          <path
            d="M0 0l5 4.998L10 0z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
